File: 39. Combination sum/can have duplicates.py
#use backtracking to get all the combinations. 
#if want only one combination then just return immediately after find the first comb

#time -
# n -> number of elements
# m -> target sum
# O(n^m)
#space = O(m^(length of combs on each stack))
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:    # type: ignore
        res = []   
                  
        def bt(target, combs, idx):
            if target == 0:
                res.append(combs.copy())
                return
            if target < 0:
                return

            for i in range(idx, len(candidates)):
                c = candidates[i]
                combs.append(c)
                bt(target - c, combs, i)
                combs.pop()
        bt(target, [], 0)
        return res
                    

File: 39. Combination sum/can have duplicates.rs
impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut combs = Vec::new();
        let mut res = Vec::new();
        Self::helper(target, 0, &candidates, &mut combs, &mut res);
        return res
    }
    pub fn helper(amount: i32, idx: usize, candidates: &Vec<i32>, combs: &mut Vec<i32>, res: &mut Vec<Vec<i32>>){
        if amount == 0{
            res.push(combs.clone());
            return;
        }
        if amount < 0{
            return;
        }

        for i in idx..candidates.len(){
            let c = candidates[i];
            combs.push(c);
            Self::helper(amount - c, i, candidates, combs, res);
            combs.pop();
        }
    }
}